pkg/standalone-utils/test/LidoRelayer.test.ts

```diff
@@ -72,383 +65,543 @@ describe('LidoRelayer', function () {
+
+    // Approve relayer by sender
+    await vault.instance.connect(sender).setRelayerApproval(sender.address, relayer.address, true);
+  });
 
-        describe('swap using stETH as an input', () => {
-          let singleSwap: SingleSwap;
+  const CHAINED_REFERENCE_PREFIX = 'ba10';
+  function toChainedReference(key: BigNumberish): BigNumber {
+    // The full padded prefix is 66 characters long, with 64 hex characters and the 0x prefix.
+    const paddedPrefix = `0x${CHAINED_REFERENCE_PREFIX}${'0'.repeat(64 - CHAINED_REFERENCE_PREFIX.length)}`;
+
+    return BigNumber.from(paddedPrefix).add(key);
+  }
+
+  function encodeApprove(token: Token, amount: BigNumberish): string {
+    return relayerLibrary.interface.encodeFunctionData('approveVault', [token.address, amount]);
+  }
+
+  function encodeWrap(sender: string, recipient: string, amount: BigNumberish, outputReference?: BigNumberish): string {
+    return relayerLibrary.interface.encodeFunctionData('wrapStETH', [sender, recipient, amount, outputReference ?? 0]);
+  }
+
+  function encodeUnwrap(
+    sender: string,
+    recipient: string,
+    amount: BigNumberish,
+    outputReference?: BigNumberish
+  ): string {
+    return relayerLibrary.interface.encodeFunctionData('unwrapWstETH', [
+      sender,
+      recipient,
+      amount,
+      outputReference ?? 0,
+    ]);
+  }
 
-          sharedBeforeEach('build swap request', async () => {
-            singleSwap = {
+  describe('swap', () => {
+    function encodeSwap(params: {
+      poolId: string;
+      kind: SwapKind;
+      tokenIn: Token;
+      tokenOut: Token;
+      amount: BigNumberish;
+      sender: Account;
+      recipient: Account;
+      outputReference?: BigNumberish;
+    }): string {
+      return relayerLibrary.interface.encodeFunctionData('swap', [
+        {
+          poolId: params.poolId,
+          kind: params.kind,
+          assetIn: params.tokenIn.address,
+          assetOut: params.tokenOut.address,
+          amount: params.amount,
+          userData: '0x',
+        },
+        {
+          sender: TypesConverter.toAddress(params.sender),
+          recipient: TypesConverter.toAddress(params.recipient),
+          fromInternalBalance: false,
+          toInternalBalance: false,
+        },
+        0,
+        MAX_UINT256,
+        0,
+        params.outputReference ?? 0,
+      ]);
+    }
+
+    describe('swap using stETH as an input', () => {
+      it('performs the given swap', async () => {
+        const poolId = basePoolId;
+        const tokenIn = tokens.findBySymbol('wstETH');
+        const tokenOut = tokens.WETH;
+        const amount = fp(1);
+
+        const receipt = await (
+          await relayer.connect(sender).multicall([
+            encodeWrap(sender.address, relayer.address, amount, toChainedReference(0)),
+            encodeApprove(tokenIn, MAX_UINT256),
+            encodeSwap({
+              poolId,
```

Comment:
Related to the comment above, now that we have chained references we could also structure this in a different way by sending the wstETH back to the sender and swapping from their wallet, avoiding the approval. This would make e.g. joining a pool with wstETH + other token easier.

Reply:
I'd say this is non ideal as it would require the user to have an approval on the vault for wstETH whereas having the relayer be the sender allows users to trade while only having a stETH approval.

Reply:
True, but it requires manually granting allowance for the relayer (maybe #991 will improve this). The `wstETH` approval would be a one-off, and the joins would be easier. As long as we support both we'll be fine.